# Case Study Model

## General Remarks

- Reverse Power Flow / Cost at substation
    - If the negative generation limit at the substation is not enforced, then substation has the option of pushing energy back towards the transmission system (reverse power flow)
    - We make the follwoing assumptions: 
        - reverse power flow is allowed at the substation 
        - for scheduled negative active power the substation is remunerated according to the market price (i.e. linear cost of energy provision at the substation) -> the substation has a linear cost function 
        - While the cost function will not have a quadratic termn, THERE WILL BE a penalty on $\alpha_0$ as it indicates how much the substation is willing to deviate from the schedule


## Notes and Todos:

- Verify the gamma decomposition
- Maybe check if 33bus gives nice/better results?
 
 
Done:   
- decide for a case! (combination of voltage limits and generators). Notes:
    - should there be an equal lambda case?
    - should the DERs be fully utilized in the deterministic case?
    - [Done]: Choosen Ex2 as set up
- put results handling in a function [done]
- remove unneccsray packages (e.g. gurobi from tomls) [done, but check on the "remains"]
- Remove quadratic term from substation [done]

In [3]:
# IJulia.qtconsole();

# set width of notbook 
display("text/html", "<style>.container { width:80% !important; }</style>")

In [4]:
# Set up environment from .toml-files
import Pkg
Pkg.activate(".")
# Pkg.instantiate()

# Load necessary packages
using DataFrames, CSV, LinearAlgebra, Dates # Basic Julia utility
using JLD, HDF5
using JuMP # Modeling Framework
using Mosek, MosekTools, Gurobi # Solver and Solver-Environment
using Logging # For some warnings in the functions

# Load functions and model
include("src/tools.jl") # Some additional functions
include("src/input.jl") # Type definitions and read-in functions
include("src/model_definition.jl") # Model definiton
include("src/output.jl") # Postprocessing of solved model

results_to_df (generic function with 1 method)

# 15Bus Case

- 15 bus case from Papavasilou paper
    - negative load is positive load
    - no line constraints
    - voltage at root is 1 p.u.
    
- Assumptions for all cases except mentioned differently
    - Two DERs one each at node 6 and 11 with `c1 = 10`, `c2 = 5`, `p_max = 1`, `q_max = 0.5`
    - No reverse power flow limit on substation
    - Substation pays `c1 = 50` for energy provision from the transmission network (no quadratic term) and a penalty of `200` on alpha to discourage high alpha

## Experiment 5

### Description

Ex5:
- tag: `experiment5` 
- all nodes have 20% standard deviation of their forecasted load
- all parameters as in AP's paper, including the negative load (i.e. uncontrollable generator with the same standard deviation
- EXCEPT: lines 2 and 3 also have s_max = 2 (instead of 0.256)
- Generation confidene: 95%, voltage confidence: 99%
- substation balancing participation penalty is set to 1000 (i.e. quadratic cost term)

Runs:
- `ex5_no_uncert`: deterministic case
- `ex5_uncert`: case with uncertainty
- `ex5_gen_uncert`: chance constraints only in generation


### Notes and Observations



In [12]:
experiment5 = Dict()

# Load feeder data 
feeder = load_feeder("data/feeder_data/basecase_noneg/")

# Set up
relative_std = 0.2
loads = [b.d_P for b in feeder.buses]
std_vec = abs.(loads) .* relative_std
var_vec = std_vec.^2
Σ = diagm(0 => var_vec)

# ex5_general_settings
ex5_general_settings = Dict(
    "var_vec" => var_vec,
    "Σ" => Σ,
    "z_g" => 1.945, # 95%
    "z_v" => 2.576, # 99%
    "z_f" => 2.576, # 99%
    "toggle_volt_cc" => false,
    "toggle_gen_cc" => false,
    "toggle_thermal_cc" => false,
    "thermal_const_method" => 2, #1=soc, 2=inner approx
    "vfac" => 0, # directly sets symetric voltage limit to min:(1-vfac) max:(1+vfac), use import values if vfac=0
    "qcfac" => 1,
    "output_level" => 0,
    "Ψ" => 0,
)

# ex5_no_uncert:
ex5_no_uncert_settings = copy(ex5_general_settings)
ex5_no_uncert_settings["toggle_volt_cc"] = false
ex5_no_uncert_settings["toggle_gen_cc"] = false
ex5_no_uncert_settings["toggle_thermal_cc"] = false

experiment5["no_uncert"] = Dict(
    "verbose" => "no_uncert",
    "feeder" => feeder,
    "settings" => ex5_no_uncert_settings,
)

# ex5_uncert:
ex5_uncert_settings = copy(ex5_general_settings)
ex5_uncert_settings["toggle_volt_cc"] = true
ex5_uncert_settings["toggle_gen_cc"] = true
ex5_uncert_settings["toggle_thermal_cc"] = false
experiment5["uncert"] = Dict(
    "verbose" => "uncert",
    "feeder" => feeder,
    "settings" => ex5_uncert_settings,
)

# ex5_gen_uncert:
ex5_gen_uncert_settings = copy(ex5_general_settings)
ex5_gen_uncert_settings["toggle_volt_cc"] = false
ex5_gen_uncert_settings["toggle_gen_cc"] = true
ex5_gen_uncert_settings["toggle_thermal_cc"] = false
experiment5["gen_uncert"] = Dict(
    "verbose" => "gen_uncert",
    "feeder" => feeder,
    "settings" => ex5_gen_uncert_settings,
)

# Run 
experiment5 = run_experiment(experiment5)

# Show results
sleep(.2) # just to ensure correct printing order
for case in keys(experiment5)
    print(">>> Case $case")
    sleep(.2)
    display(experiment5[case]["results"])
end

>>>>> Reading feeder data from data/feeder_data/basecase_noneg/
Running uncert
>>>> Building Model
>>>> Running Model
MOSEK warning 705: #3 (nearly) zero elements are specified in sparse row ''(86) of matrix 'A'.
MOSEK warning 705: #4 (nearly) zero elements are specified in sparse row ''(92) of matrix 'A'.
>>>> Model finished with status OPTIMAL in 0.037558884 seconds
>>>> Post-Processing

Running gen_uncert
>>>> Building Model
>>>> Running Model
>>>> Model finished with status OPTIMAL in 0.019324034 seconds
>>>> Post-Processing

Running no_uncert
>>>> Building Model
>>>> Running Model
>>>> Model finished with status OPTIMAL in 0.009955943 seconds
>>>> Post-Processing

>>> Case uncert

15×28 DataFrame. Omitted printing of 22 columns
│ Row │ objective │ bus │ gp       │ gq         │ voltage  │ fp         │
│     │ Float64   │ Any │ Any      │ Any        │ Any      │ Any        │
├─────┼───────────┼─────┼──────────┼────────────┼──────────┼────────────┤
│ 1   │ 71.7329   │ 1   │ 1.1981   │ 0.378944   │ 1.0      │ 0          │
│ 2   │ 0.0       │ 2   │ 0        │ 0          │ 0.970112 │ 0.5524     │
│ 3   │ 0.0       │ 3   │ 0        │ 0          │ 0.984769 │ -0.2412    │
│ 4   │ 0.0       │ 4   │ 0        │ 0          │ 1.00731  │ -0.2412    │
│ 5   │ 0.0       │ 5   │ 0        │ 0          │ 1.00937  │ -0.0372927 │
│ 6   │ 0.0       │ 6   │ 0        │ 0          │ 1.01166  │ -0.0545927 │
│ 7   │ 0.0       │ 7   │ 0.105593 │ 0.0669092  │ 1.01979  │ -0.0836927 │
│ 8   │ 0.0       │ 8   │ 0        │ 0          │ 1.02077  │ -0.1969    │
│ 9   │ 0.0       │ 9   │ 0        │ 0          │ 1.01077  │ -0.224007  │
│ 10  │ 0.0       │ 10  │ 0        │ 0          │ 1.01002  │ -0.0506074 │
│ 11  │ 0.0       │ 11  │ 0        │ 0          │ 1.00922  │ -0.0735074 │
│ 12  │ 0.0       │ 12  │ 0.108407 │ -0.0648531 │ 1.00919  │ -0.0952074 │
│ 13  │ 0.0       │ 13  │ 0        │ 0          │ 0.982619 │ 0.6457     │
│ 14  │ 0.0       │ 14  │ 0        │ 0          │ 0.977795 │ 0.0238     │
│ 15  │ 0.0       │ 15  │ 0        │ 0          │ 0.975027 │ 0.0224     │

>>> Case gen_uncert

15×28 DataFrame. Omitted printing of 22 columns
│ Row │ objective │ bus │ gp       │ gq        │ voltage  │ fp         │
│     │ Float64   │ Any │ Any      │ Any       │ Any      │ Any        │
├─────┼───────────┼─────┼──────────┼───────────┼──────────┼────────────┤
│ 1   │ 70.4343   │ 1   │ 1.1833   │ 0.3237    │ 1.0      │ 0          │
│ 2   │ 0.0       │ 2   │ 0        │ 0         │ 0.976936 │ 0.5376     │
│ 3   │ 0.0       │ 3   │ 0        │ 0         │ 0.999807 │ -0.256     │
│ 4   │ 0.0       │ 4   │ 0        │ 0         │ 1.03464  │ -0.256     │
│ 5   │ 0.0       │ 5   │ 0        │ 0         │ 1.03576  │ -0.0461079 │
│ 6   │ 0.0       │ 6   │ 0        │ 0         │ 1.03718  │ -0.0634079 │
│ 7   │ 0.0       │ 7   │ 0.114408 │ 0.0273331 │ 1.04291  │ -0.0925079 │
│ 8   │ 0.0       │ 8   │ 0        │ 0         │ 1.05323  │ -0.1969    │
│ 9   │ 0.0       │ 9   │ 0        │ 0         │ 1.04354  │ -0.229992  │
│ 10  │ 0.0       │ 10  │ 0        │ 0         │ 1.04417  │ -0.0565921 │
│ 11  │ 0.0       │ 11  │ 0        │ 0         │ 1.04667  │ -0.0794921 │
│ 12  │ 0.0       │ 12  │ 0.114392 │ 0.0299669 │ 1.04804  │ -0.101192  │
│ 13  │ 0.0       │ 13  │ 0        │ 0         │ 0.982619 │ 0.6457     │
│ 14  │ 0.0       │ 14  │ 0        │ 0         │ 0.977795 │ 0.0238     │
│ 15  │ 0.0       │ 15  │ 0        │ 0         │ 0.975027 │ 0.0224     │

>>> Case no_uncert

15×28 DataFrame. Omitted printing of 22 columns
│ Row │ objective │ bus │ gp       │ gq       │ voltage  │ fp         │
│     │ Float64   │ Any │ Any      │ Any      │ Any      │ Any        │
├─────┼───────────┼─────┼──────────┼──────────┼──────────┼────────────┤
│ 1   │ 61.5275   │ 1   │ 1.1833   │ 0.3237   │ 1.0      │ 0          │
│ 2   │ 0.0       │ 2   │ 0        │ 0        │ 0.976936 │ 0.5376     │
│ 3   │ 0.0       │ 3   │ 0        │ 0        │ 0.999807 │ -0.256     │
│ 4   │ 0.0       │ 4   │ 0        │ 0        │ 1.03464  │ -0.256     │
│ 5   │ 0.0       │ 5   │ 0        │ 0        │ 1.0357   │ -0.0460628 │
│ 6   │ 0.0       │ 6   │ 0        │ 0        │ 1.03708  │ -0.0633628 │
│ 7   │ 0.0       │ 7   │ 0.114363 │ 0.025303 │ 1.04267  │ -0.0924628 │
│ 8   │ 0.0       │ 8   │ 0        │ 0        │ 1.05334  │ -0.1969    │
│ 9   │ 0.0       │ 9   │ 0        │ 0        │ 1.04366  │ -0.230037  │
│ 10  │ 0.0       │ 10  │ 0        │ 0        │ 1.04431  │ -0.0566372 │
│ 11  │ 0.0       │ 11  │ 0        │ 0        │ 1.04688  │ -0.0795372 │
│ 12  │ 0.0       │ 12  │ 0.114437 │ 0.031997 │ 1.04828  │ -0.101237  │
│ 13  │ 0.0       │ 13  │ 0        │ 0        │ 0.982619 │ 0.6457     │
│ 14  │ 0.0       │ 14  │ 0        │ 0        │ 0.977795 │ 0.0238     │
│ 15  │ 0.0       │ 15  │ 0        │ 0        │ 0.975027 │ 0.0224     │

In [11]:
save_experiment("experiment5", experiment5)

Saving uncert
Saving gen_uncert
Saving no_uncert
>>>> Saved with timestamp 190422_1211


In [8]:
m, meta = build_model(experiment5["uncert"]["feeder"], experiment5["uncert"]["settings"])
optimize!(m)

MOSEK warning 705: #3 (nearly) zero elements are specified in sparse row ''(86) of matrix 'A'.
MOSEK warning 705: #4 (nearly) zero elements are specified in sparse row ''(92) of matrix 'A'.


In [9]:
A = zeros(14, 12)
for b in 2:15
    for c in 1:12
        A[b-1, c] = shadow_price(m[:η][b,c])
    end
end

DataFrame(A)

14×12 DataFrame. Omitted printing of 7 columns
│ Row │ x1          │ x2          │ x3          │ x4          │ x5          │
│     │ Float64     │ Float64     │ Float64     │ Float64     │ Float64     │
├─────┼─────────────┼─────────────┼─────────────┼─────────────┼─────────────┤
│ 1   │ -1.09382e-6 │ -7.92339e-7 │ -1.06672e-6 │ 0.0         │ 0.0         │
│ 2   │ 0.0         │ 0.0         │ 0.0         │ 0.0         │ -6.94047e-7 │
│ 3   │ 0.0         │ 0.0         │ 0.0         │ -9.31583e-7 │ 0.0         │
│ 4   │ -4.16429e-6 │ -3.9917e-6  │ -8.32058e-6 │ 0.0         │ 0.0         │
│ 5   │ 0.0         │ 0.0         │ 0.0         │ 0.0         │ 0.0         │
│ 6   │ 0.0         │ 0.0         │ 0.0         │ -1.5629e-5  │ -4.75999e-5 │
│ 7   │ 0.0         │ 0.0         │ 0.0         │ 0.0         │ 0.0         │
│ 8   │ 0.0         │ 0.0         │ 0.0         │ -1.4641e-5  │ -4.10641e-5 │
│ 9   │ -5.83368e-6 │ -5.09615e-6 │ -8.52663e-6 │ 0.0         │ 0.0         │
│ 10  │ 0.0         │ 0.0         │ 0.0         │ 0.0         │ 0.0         │
│ 11  │ 0.0         │ 0.0         │ 0.0         │ 0.0         │ 0.0         │
│ 12  │ 0.0         │ 0.0         │ 0.0         │ 0.0         │ 0.0         │
│ 13  │ 0.0         │ 0.0         │ 0.0         │ 0.0         │ 0.0         │
│ 14  │ 0.0         │ 0.0         │ 0.0         │ 0.0         │ 0.0         │

In [18]:
fed = experiment5["uncert"]["feeder"]
traverse(fed, 12)

1-element Array{Any,1}:
 12

## Experiment 1

We analyse the influence of the uncertainty in the system using the "standard" setup

### Description

Ex1:
- tag: `experiment1`
- tight bascase voltage (0.95-1.05) 
- all nodes have 20% standard deviation of their forecasted load

Runs:
- `ex1_no_uncert`: deterministic case
- `ex1_uncert`: case with uncertainty
- `ex1_gen_uncert`: Chance constraints only in generation


### Notes and Observations

- Recursive decomposition of prices for active power (eqtns (20) and (21)) are verified by the experimet
- Decompositon via voltage constraints (eqtns (23) and (24)) also verified by the experiment

In [36]:
experiment1 = Dict()

# Load feeder data 
feeder = load_feeder("data/feeder_data/basecase_lv_noneg/")

# Set up
relative_std = 0.2
loads = [b.d_P for b in feeder.buses]
std_vec = loads .* relative_std
var_vec = std_vec.^2
Σ = diagm(0 => var_vec)

# ex1_general_settings
ex1_general_settings = Dict(
    "var_vec" => var_vec,
    "Σ" => Σ,
    "z_g" => 1.945,
    "z_v" => 2.576,
    "toggle_volt_cc" => false,
    "toggle_gen_cc" => false,
    "vfac" => 0, # directly sets symetric voltage limit to min:(1-vfac) max:(1+vfac), use import values if vfac=0
    "qcfac" => 1,
    "output_level" => 0,
    "Ψ" => 0,
)


# ex1_no_uncert:
ex1_no_uncert_settings = copy(ex1_general_settings)
ex1_no_uncert_settings["toggle_volt_cc"] = false
ex1_no_uncert_settings["toggle_gen_cc"] = false

experiment1["no_uncert"] = Dict(
    "verbose" => "ex1_no_uncert",
    "feeder" => feeder,
    "settings" => ex1_no_uncert_settings,
)

# ex1_uncert:
ex1_uncert_settings = copy(ex1_general_settings)
ex1_uncert_settings["toggle_volt_cc"] = true
ex1_uncert_settings["toggle_gen_cc"] = true
experiment1["uncert"] = Dict(
    "verbose" => "ex1_uncert",
    "feeder" => feeder,
    "settings" => ex1_uncert_settings,
)

# ex1_gen_uncert:
ex1_uncert_settings = copy(ex1_general_settings)
ex1_uncert_settings["toggle_volt_cc"] = false
ex1_uncert_settings["toggle_gen_cc"] = true
experiment1["gen_uncert"] = Dict(
    "verbose" => "ex1_gen_uncert",
    "feeder" => feeder,
    "settings" => ex1_uncert_settings,
)

# Run 
experiment1 = run_experiment(experiment1)

# Show results
sleep(.2) # just to ensure correct printing order
for case in keys(experiment1)
    display(experiment1[case]["results"])
end

>>>>> Reading feeder data from data/feeder_data/basecase_lv_noneg/
Running ex1_uncert
>>>> Building Model


KeyError: KeyError: key "toggle_thermal" not found

In [3]:
save_experiment("experiment1", experiment1)

Saving ex1_uncert
Saving ex1_gen_uncert
Saving ex1_no_uncert
>>>> Saved with timestamp 190320_0927


## Experiment 2

Same as Experiment1 but with wider voltage constraints

### Description

Ex2:
- tag: `experiment2` 
- bascase voltage (0.9-1.1) and `vfac = 0.02`: Resulting "real" voltage limits: ca. 0.91 - 1.09 (i.e. +- 9%)
- all nodes have 20% standard deviation of their forecasted load
- all parameters as in AP's paper, including the negative load (i.e. uncontrollable generator with the same standard deviation
- substation balancing participation penalty is set to 1000 (i.e. quadratic cost term)

Runs:
- `ex2_no_uncert`: deterministic case
- `ex2_uncert`: case with uncertainty
- `ex2_gen_uncert`: chance constraints only in generation


### Notes and Observations

- Without uncertainty the network is "uncongested" so that price for active power is the same at all nodes
- With uncertainty, some voltage constraints become binding and there are differences in the prices

In [8]:
experiment2 = Dict()

# Load feeder data 
feeder = load_feeder("data/feeder_data/basecase_lv_noneg/")

# Set up
relative_std = 0.2
loads = [b.d_P for b in feeder.buses]
std_vec = abs.(loads) .* relative_std
var_vec = std_vec.^2
Σ = diagm(0 => var_vec)

# ex2_general_settings
ex2_general_settings = Dict(
    "var_vec" => var_vec,
    "Σ" => Σ,
    "z_g" => 1.945,
    "z_v" => 2.576,
    "toggle_volt_cc" => false,
    "toggle_gen_cc" => false,
    "toggle_thermal_cc" => false,
    "thermal_const_method" => 0, #1=soc, 2=inner approx, other=off
    "vfac" => 0, # directly sets symetric voltage limit to min:(1-vfac) max:(1+vfac), use import values if vfac=0
    "qcfac" => 1,
    "output_level" => 0,
    "Ψ" => 0,
)

# ex2_no_uncert:
ex2_no_uncert_settings = copy(ex2_general_settings)
ex2_no_uncert_settings["toggle_volt_cc"] = false
ex2_no_uncert_settings["toggle_gen_cc"] = false

experiment2["no_uncert"] = Dict(
    "verbose" => "no_uncert",
    "feeder" => feeder,
    "settings" => ex2_no_uncert_settings,
)

# ex2_uncert:
ex2_uncert_settings = copy(ex2_general_settings)
ex2_uncert_settings["toggle_volt_cc"] = true
ex2_uncert_settings["toggle_gen_cc"] = true
experiment2["uncert"] = Dict(
    "verbose" => "uncert",
    "feeder" => feeder,
    "settings" => ex2_uncert_settings,
)

# ex2_gen_uncert:
ex2_uncert_settings = copy(ex2_general_settings)
ex2_uncert_settings["toggle_volt_cc"] = false
ex2_uncert_settings["toggle_gen_cc"] = true
experiment2["gen_uncert"] = Dict(
    "verbose" => "gen_uncert",
    "feeder" => feeder,
    "settings" => ex2_uncert_settings,
)

# Run 
experiment2 = run_experiment(experiment2)

# Show results
sleep(.2) # just to ensure correct printing order
for case in keys(experiment2)
    print(">>> Case $case")
    sleep(.2)
    display(experiment2[case]["results"])
end

>>>>> Reading feeder data from data/feeder_data/basecase_lv_noneg/
Running uncert
>>>> Building Model
>>>> Running Model
MOSEK warning 705: #3 (nearly) zero elements are specified in sparse row ''(86) of matrix 'A'.
MOSEK warning 705: #4 (nearly) zero elements are specified in sparse row ''(92) of matrix 'A'.
>>>> Model finished with status OPTIMAL in 0.020427029 seconds
>>>> Post-Processing


┌ Warning: Thermal constraint method 0 unknown. Proceeding with unconstrained lines.
└ @ Main C:\Users\Robert\Promotion\Workshop\02_cc_dlmp_varaware\01_casestudy_uncertan_dlmp\src\model_definition.jl:196



Running gen_uncert
>>>> Building Model
>>>> Running Model
>>>> Model finished with status OPTIMAL in 0.003514463 seconds
>>>> Post-Processing

Running no_uncert
>>>> Building Model
>>>> Running Model
>>>> Model finished with status OPTIMAL in 0.002372737 seconds
>>>> Post-Processing



┌ Warning: Thermal constraint method 0 unknown. Proceeding with unconstrained lines.
└ @ Main C:\Users\Robert\Promotion\Workshop\02_cc_dlmp_varaware\01_casestudy_uncertan_dlmp\src\model_definition.jl:196
┌ Warning: Thermal constraint method 0 unknown. Proceeding with unconstrained lines.
└ @ Main C:\Users\Robert\Promotion\Workshop\02_cc_dlmp_varaware\01_casestudy_uncertan_dlmp\src\model_definition.jl:196


>>> Case uncert

15×23 DataFrame. Omitted printing of 17 columns
│ Row │ objective │ bus │ gp       │ gq         │ voltage  │ fp        │
│     │ Float64   │ Any │ Any      │ Any        │ Any      │ Any       │
├─────┼───────────┼─────┼──────────┼────────────┼──────────┼───────────┤
│ 1   │ 68.8929   │ 1   │ 0.9225   │ 0.538211   │ 1.0      │ 0         │
│ 2   │ 0.0       │ 2   │ 0        │ 0          │ 0.950497 │ -0.0      │
│ 3   │ 0.0       │ 3   │ 0        │ 0          │ 0.969829 │ 0.2768    │
│ 4   │ 0.0       │ 4   │ 0        │ 0          │ 0.999378 │ -0.5168   │
│ 5   │ 0.0       │ 5   │ 0        │ 0          │ 1.00313  │ -0.5168   │
│ 6   │ 0.0       │ 6   │ 0        │ 0          │ 1.00696  │ -0.372884 │
│ 7   │ 0.0       │ 7   │ 0.441184 │ -0.106042  │ 1.01932  │ -0.390184 │
│ 8   │ 0.0       │ 8   │ 0        │ 0          │ 0.990743 │ -0.419284 │
│ 9   │ 0.0       │ 9   │ 0        │ 0          │ 1.00122  │ 0.1969    │
│ 10  │ 0.0       │ 10  │ 0        │ 0          │ 1.00399  │ -0.164016 │
│ 11  │ 0.0       │ 11  │ 0        │ 0          │ 1.01164  │ -0.384416 │
│ 12  │ 0.0       │ 12  │ 0.442216 │ -0.0511695 │ 1.0152   │ -0.407316 │
│ 13  │ 0.0       │ 13  │ 0        │ 0          │ 0.982619 │ -0.429016 │
│ 14  │ 0.0       │ 14  │ 0        │ 0          │ 0.977795 │ 0.6457    │
│ 15  │ 0.0       │ 15  │ 0        │ 0          │ 0.975027 │ 0.0238    │

>>> Case gen_uncert

15×23 DataFrame. Omitted printing of 17 columns
│ Row │ objective │ bus │ gp       │ gq         │ voltage  │ fp        │
│     │ Float64   │ Any │ Any      │ Any        │ Any      │ Any       │
├─────┼───────────┼─────┼──────────┼────────────┼──────────┼───────────┤
│ 1   │ 52.7535   │ 1   │ 0.787837 │ 0.543266   │ 1.0      │ 0         │
│ 2   │ 0.0       │ 2   │ 0        │ 0          │ 0.95     │ 0.142137  │
│ 3   │ 0.0       │ 3   │ 0        │ 0          │ 0.980882 │ -0.651463 │
│ 4   │ 0.0       │ 4   │ 0        │ 0          │ 1.02742  │ -0.651463 │
│ 5   │ 0.0       │ 5   │ 0        │ 0          │ 1.0318   │ -0.440129 │
│ 6   │ 0.0       │ 6   │ 0        │ 0          │ 1.03618  │ -0.457429 │
│ 7   │ 0.0       │ 7   │ 0.508429 │ -0.12566   │ 1.05     │ -0.486529 │
│ 8   │ 0.0       │ 8   │ 0        │ 0          │ 1.02254  │ 0.1969    │
│ 9   │ 0.0       │ 9   │ 0        │ 0          │ 1.0327   │ -0.231434 │
│ 10  │ 0.0       │ 10  │ 0        │ 0          │ 1.03623  │ -0.451834 │
│ 11  │ 0.0       │ 11  │ 0        │ 0          │ 1.04568  │ -0.474734 │
│ 12  │ 0.0       │ 12  │ 0.509634 │ -0.0366061 │ 1.05     │ -0.496434 │
│ 13  │ 0.0       │ 13  │ 0        │ 0          │ 0.982619 │ 0.6457    │
│ 14  │ 0.0       │ 14  │ 0        │ 0          │ 0.977795 │ 0.0238    │
│ 15  │ 0.0       │ 15  │ 0        │ 0          │ 0.975027 │ 0.0224    │

>>> Case no_uncert

15×23 DataFrame. Omitted printing of 17 columns
│ Row │ objective │ bus │ gp       │ gq         │ voltage  │ fp        │
│     │ Float64   │ Any │ Any      │ Any        │ Any      │ Any       │
├─────┼───────────┼─────┼──────────┼────────────┼──────────┼───────────┤
│ 1   │ 49.9039   │ 1   │ 0.787837 │ 0.543266   │ 1.0      │ 0         │
│ 2   │ 0.0       │ 2   │ 0        │ 0          │ 0.95     │ 0.142137  │
│ 3   │ 0.0       │ 3   │ 0        │ 0          │ 0.980882 │ -0.651463 │
│ 4   │ 0.0       │ 4   │ 0        │ 0          │ 1.02742  │ -0.651463 │
│ 5   │ 0.0       │ 5   │ 0        │ 0          │ 1.0318   │ -0.438523 │
│ 6   │ 0.0       │ 6   │ 0        │ 0          │ 1.03618  │ -0.455823 │
│ 7   │ 0.0       │ 7   │ 0.506823 │ -0.124537  │ 1.05     │ -0.484923 │
│ 8   │ 0.0       │ 8   │ 0        │ 0          │ 1.02254  │ 0.1969    │
│ 9   │ 0.0       │ 9   │ 0        │ 0          │ 1.0327   │ -0.23304  │
│ 10  │ 0.0       │ 10  │ 0        │ 0          │ 1.03623  │ -0.45344  │
│ 11  │ 0.0       │ 11  │ 0        │ 0          │ 1.04568  │ -0.47634  │
│ 12  │ 0.0       │ 12  │ 0.51124  │ -0.0377283 │ 1.05     │ -0.49804  │
│ 13  │ 0.0       │ 13  │ 0        │ 0          │ 0.982619 │ 0.6457    │
│ 14  │ 0.0       │ 14  │ 0        │ 0          │ 0.977795 │ 0.0238    │
│ 15  │ 0.0       │ 15  │ 0        │ 0          │ 0.975027 │ 0.0224    │

In [44]:
save_experiment("experiment2", experiment2)

Saving uncert
Saving gen_uncert
Saving no_uncert
>>>> Saved with timestamp 190320_1040


## Experiment 3

Same as before but instead of tighter voltage limits try higher load

In [10]:
experiment3 = Dict()

# Load feeder data 
feeder = load_feeder("data/feeder_data/basecase_noneg/")

# Set up
relative_std = 0.2
loads = [b.d_P for b in feeder.buses]
std_vec = loads .* relative_std
var_vec = std_vec.^2
Σ = diagm(0 => var_vec)

# ex3_general_settings
ex3_general_settings = Dict(
    "var_vec" => var_vec,
    "Σ" => Σ,
    "z_g" => 1.945,
    "z_v" => 2.576,
    "toggle_volt_cc" => false,
    "toggle_gen_cc" => false,
    "vfac" => 0, # decrease upper bound by factor (1-vfac) and increase lower bound by (1+vfac)
    "qcfac" => 1,
    "output_level" => 0,
    "Ψ" => 0,
    "loadfac" => 3
)

# ex3_no_uncert:
ex3_no_uncert_settings = copy(ex3_general_settings)
ex3_no_uncert_settings["toggle_volt_cc"] = false
ex3_no_uncert_settings["toggle_gen_cc"] = false

experiment3["no_uncert"] = Dict(
    "verbose" => "ex3_no_uncert",
    "feeder" => feeder,
    "settings" => ex3_no_uncert_settings,
)

# ex3_uncert:
ex3_uncert_settings = copy(ex3_general_settings)
ex3_uncert_settings["toggle_volt_cc"] = true
ex3_uncert_settings["toggle_gen_cc"] = true
experiment3["uncert"] = Dict(
    "verbose" => "ex3_uncert",
    "feeder" => feeder,
    "settings" => ex3_uncert_settings,
)

# ex3_gen_uncert:
ex3_uncert_settings = copy(ex3_general_settings)
ex3_uncert_settings["toggle_volt_cc"] = false
ex3_uncert_settings["toggle_gen_cc"] = true
experiment3["gen_uncert"] = Dict(
    "verbose" => "ex3_gen_uncert",
    "feeder" => feeder,
    "settings" => ex3_uncert_settings,
)

# Run 
experiment3 = run_experiment(experiment3)

# Show results
sleep(.2) # just to ensure correct printing order
for case in keys(experiment3)
    print(">>> Case $case")
    sleep(.2)
    display(experiment3[case]["results"])
end

>>>>> Reading feeder data from data/feeder_data/basecase_noneg/
Running ex3_uncert
>>>> Building Model
>>>> Running Model
MOSEK warning 705: #3 (nearly) zero elements are specified in sparse row ''(87) of matrix 'A'.
MOSEK warning 705: #4 (nearly) zero elements are specified in sparse row ''(93) of matrix 'A'.
>>>> Model finished with status OPTIMAL in 0.006354567 seconds
>>>> Post-Processing

Running ex3_gen_uncert
>>>> Building Model
>>>> Running Model
>>>> Model finished with status OPTIMAL in 0.002476817 seconds
>>>> Post-Processing

Running ex3_no_uncert
>>>> Building Model
>>>> Running Model
>>>> Model finished with status OPTIMAL in 0.00223549 seconds
>>>> Post-Processing

>>> Case uncert

15×19 DataFrame. Omitted printing of 12 columns
│ Row │ objective │ bus │ gp     │ gq         │ voltage  │ alpha │ lambda   │
│     │ Float64   │ Any │ Any    │ Any        │ Any      │ Any   │ Any      │
├─────┼───────────┼─────┼────────┼────────────┼──────────┼───────┼──────────┤
│ 1   │ 431.923   │ 1   │ 3.8177 │ 1.0676     │ 1.0      │ 1.0   │ -100.0   │
│ 2   │ 0.0       │ 2   │ 0      │ 0          │ 0.916252 │ 0     │ -100.0   │
│ 3   │ 0.0       │ 3   │ 0      │ 0          │ 0.950524 │ 0     │ -100.0   │
│ 4   │ 0.0       │ 4   │ 0      │ 0          │ 1.00189  │ 0     │ -99.9999 │
│ 5   │ 0.0       │ 5   │ 0      │ 0          │ 1.0105   │ 0     │ -99.9999 │
│ 6   │ 0.0       │ 6   │ 0      │ 0          │ 1.01952  │ 0     │ -99.9999 │
│ 7   │ 0.0       │ 7   │ 0.8    │ -0.0477213 │ 1.04946  │ 0     │ -99.9999 │
│ 8   │ 0.0       │ 8   │ 0      │ 0          │ 0.970333 │ 0     │ -99.9999 │
│ 9   │ 0.0       │ 9   │ 0      │ 0          │ 1.00209  │ 0     │ -99.9999 │
│ 10  │ 0.0       │ 10  │ 0      │ 0          │ 1.00905  │ 0     │ -99.9999 │
│ 11  │ 0.0       │ 11  │ 0      │ 0          │ 1.02867  │ 0     │ -99.9999 │
│ 12  │ 0.0       │ 12  │ 0.8    │ 0.123118   │ 1.03787  │ 0     │ -99.9999 │
│ 13  │ 0.0       │ 13  │ 0      │ 0          │ 0.946901 │ 0     │ -100.0   │
│ 14  │ 0.0       │ 14  │ 0      │ 0          │ 0.931799 │ 0     │ -100.0   │
│ 15  │ 0.0       │ 15  │ 0      │ 0          │ 0.923057 │ 0     │ -100.0   │

>>> Case gen_uncert

15×19 DataFrame. Omitted printing of 12 columns
│ Row │ objective │ bus │ gp     │ gq          │ voltage  │ alpha │ lambda │
│     │ Float64   │ Any │ Any    │ Any         │ Any      │ Any   │ Any    │
├─────┼───────────┼─────┼────────┼─────────────┼──────────┼───────┼────────┤
│ 1   │ 431.923   │ 1   │ 3.8177 │ 1.06751     │ 1.0      │ 1.0   │ -100.0 │
│ 2   │ 0.0       │ 2   │ 0      │ 0           │ 0.916265 │ 0     │ -100.0 │
│ 3   │ 0.0       │ 3   │ 0      │ 0           │ 0.950548 │ 0     │ -100.0 │
│ 4   │ 0.0       │ 4   │ 0      │ 0           │ 1.00193  │ 0     │ -100.0 │
│ 5   │ 0.0       │ 5   │ 0      │ 0           │ 1.01167  │ 0     │ -100.0 │
│ 6   │ 0.0       │ 6   │ 0      │ 0           │ 1.02172  │ 0     │ -100.0 │
│ 7   │ 0.0       │ 7   │ 0.8    │ -0.00573882 │ 1.05434  │ 0     │ -100.0 │
│ 8   │ 0.0       │ 8   │ 0      │ 0           │ 0.96786  │ 0     │ -100.0 │
│ 9   │ 0.0       │ 9   │ 0      │ 0           │ 0.999696 │ 0     │ -100.0 │
│ 10  │ 0.0       │ 10  │ 0      │ 0           │ 1.00608  │ 0     │ -100.0 │
│ 11  │ 0.0       │ 11  │ 0      │ 0           │ 1.02434  │ 0     │ -100.0 │
│ 12  │ 0.0       │ 12  │ 0.8    │ 0.0812299   │ 1.03298  │ 0     │ -100.0 │
│ 13  │ 0.0       │ 13  │ 0      │ 0           │ 0.946901 │ 0     │ -100.0 │
│ 14  │ 0.0       │ 14  │ 0      │ 0           │ 0.931799 │ 0     │ -100.0 │
│ 15  │ 0.0       │ 15  │ 0      │ 0           │ 0.923057 │ 0     │ -100.0 │

>>> Case no_uncert

15×19 DataFrame. Omitted printing of 12 columns
│ Row │ objective │ bus │ gp     │ gq          │ voltage  │ alpha │ lambda │
│     │ Float64   │ Any │ Any    │ Any         │ Any      │ Any   │ Any    │
├─────┼───────────┼─────┼────────┼─────────────┼──────────┼───────┼────────┤
│ 1   │ 399.29    │ 1   │ 3.8177 │ 1.07126     │ 1.0      │ 0     │ -100.0 │
│ 2   │ 0.0       │ 2   │ 0      │ 0           │ 0.915774 │ 0     │ -100.0 │
│ 3   │ 0.0       │ 3   │ 0      │ 0           │ 0.949578 │ 0     │ -100.0 │
│ 4   │ 0.0       │ 4   │ 0      │ 0           │ 1.00027  │ 0     │ -100.0 │
│ 5   │ 0.0       │ 5   │ 0      │ 0           │ 1.01009  │ 0     │ -100.0 │
│ 6   │ 0.0       │ 6   │ 0      │ 0           │ 1.02022  │ 0     │ -100.0 │
│ 7   │ 0.0       │ 7   │ 0.8    │ -0.00323173 │ 1.05305  │ 0     │ -100.0 │
│ 8   │ 0.0       │ 8   │ 0      │ 0           │ 0.965763 │ 0     │ -100.0 │
│ 9   │ 0.0       │ 9   │ 0      │ 0           │ 0.997666 │ 0     │ -100.0 │
│ 10  │ 0.0       │ 10  │ 0      │ 0           │ 1.00397  │ 0     │ -100.0 │
│ 11  │ 0.0       │ 11  │ 0      │ 0           │ 1.02206  │ 0     │ -100.0 │
│ 12  │ 0.0       │ 12  │ 0.8    │ 0.0749762   │ 1.03063  │ 0     │ -100.0 │
│ 13  │ 0.0       │ 13  │ 0      │ 0           │ 0.946901 │ 0     │ -100.0 │
│ 14  │ 0.0       │ 14  │ 0      │ 0           │ 0.931799 │ 0     │ -100.0 │
│ 15  │ 0.0       │ 15  │ 0      │ 0           │ 0.923057 │ 0     │ -100.0 │

## Experiment 4

- Instead of uncertain load uncontrollable vs. controllable DERs
- Maybe better in the 33-bus system?

## Prototyping Space

In [5]:
res = experiment2["uncert"]["results"]

15×20 DataFrame. Omitted printing of 14 columns
│ Row │ objective │ bus │ gp         │ gq        │ voltage  │ alpha     │
│     │ Float64   │ Any │ Any        │ Any       │ Any      │ Any       │
├─────┼───────────┼─────┼────────────┼───────────┼──────────┼───────────┤
│ 1   │ 42.1519   │ 1   │ -0.0653151 │ 0.933284  │ 1.0      │ 0.895129  │
│ 2   │ 0.0       │ 2   │ 0          │ 0         │ 0.900334 │ 0         │
│ 3   │ 0.0       │ 3   │ 0          │ 0         │ 0.960447 │ 0         │
│ 4   │ 0.0       │ 4   │ 0          │ 0         │ 1.04775  │ 0         │
│ 5   │ 0.0       │ 5   │ 0          │ 0         │ 1.05393  │ 0         │
│ 6   │ 0.0       │ 6   │ 0          │ 0         │ 1.05993  │ 0         │
│ 7   │ 0.0       │ 7   │ 0.753981   │ -0.224743 │ 1.0781   │ 0.0393694 │
│ 8   │ 0.0       │ 8   │ 0          │ 0         │ 1.06984  │ 0         │
│ 9   │ 0.0       │ 9   │ 0          │ 0         │ 1.06031  │ 0         │
│ 10  │ 0.0       │ 10  │ 0          │ 0         │ 1.06184  │ 0         │
│ 11  │ 0.0       │ 11  │ 0          │ 0         │ 1.0665   │ 0         │
│ 12  │ 0.0       │ 12  │ 0.723434   │ -0.327542 │ 1.06877  │ 0.0655016 │
│ 13  │ 0.0       │ 13  │ 0          │ 0         │ 0.982619 │ 0         │
│ 14  │ 0.0       │ 14  │ 0          │ 0         │ 0.977795 │ 0         │
│ 15  │ 0.0       │ 15  │ 0          │ 0         │ 0.975027 │ 0         │

In [1]:
0.0005*50

0.025